### PR TITLE
Autoloading, dependency loading and middleware registry cleanup

### PR DESCRIPTION
**Pull request by iMacTia**
## Description

* Remove autoloading
* Remove adapters from list of dependencies, except for net_http (doesn't require ext libraries and it's good to have a default)
* Remove dynamic dependency loader
* Simplify `MiddlewareRegistry`

## Todos
List any remaining work that needs to be done, i.e:
- [x] Tests
- [x] Upgrading
- [ ] Documentation?

### CODE REVIEW

File: UPGRADING.md [hunk 1]
```diff
@@ -31,9 +31,25 @@ We did our best to make this transition as painless as possible for you, so here
 * We've setup an [Awesome Faraday](https://github.com/lostisland/awesome-faraday) repository, where you can find and discover adapters.
   We also highlighted their unique features and level of compliance with Faraday's features.
 
+### Autoloading and dependencies
+
+Faraday have until now provided and relied on a complex dynamic dependencies system.
```
Comment (olleolleolle) [inline UPGRADING.md:36]:
```suggestion
Faraday has until now provided and relied on a complex dynamic dependencies system.
```

File: UPGRADING.md [hunk 2]
```diff
@@ -31,9 +31,25 @@ We did our best to make this transition as painless as possible for you, so here
...
+This would allow to reference classes and require dependencies only when needed (e.g. when running the first request) based
+on the middleware/adapters used.
+As part of Faraday v2.0, we've removed all external dependencies, which means we don't really need this anymore.
+This change should not affect you directly, but if you're registering middleware then beware of the new syntax:
```
Comment (olleolleolle) [inline UPGRADING.md:40]:
```suggestion
This change should not affect you directly, but if you're registering middleware then be aware of the new syntax:
```

File: UPGRADING.md [hunk 3]
```diff
@@ -31,9 +31,25 @@ We did our best to make this transition as painless as possible for you, so here
...
+
+```ruby
+# `name` here can be anything you want.
+# `klass` is your custom middleware class.
+# This method can be also called on `Faraday::Adapter`, `Faraday::Request` and `Faraday::Response`
```
Comment (olleolleolle) [inline UPGRADING.md:45]:
```suggestion
# This method can also be called on `Faraday::Adapter`, `Faraday::Request` and `Faraday::Response`
```

File: faraday.gemspec [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
 require_relative 'lib/faraday/version'
 
-Gem::Specification.new do |spec| # rubocop:disable Metrics/BlockLength
+Gem::Specification.new do |spec|
```
Comment (olleolleolle) [inline faraday.gemspec:5]: 💯

File: lib/faraday.rb [hunk 5]
```diff
@@ -169,6 +149,4 @@ def method_missing(name, *args, &block)
   self.root_path = File.expand_path __dir__
   self.lib_path = File.expand_path 'faraday', __dir__
   self.default_adapter = :net_http
-
-  require_lib 'autoload' unless ENV['FARADAY_NO_AUTOLOAD']
```
Comment (olleolleolle) [inline lib/faraday.rb:173]: Is this ENV var used anywhere in the code or CI or anything? Was it just an undocumented extra?

**Review by olleolleolle — COMMENTED** [review]
Looks grand! 

I was curious if the middleware_mutex is still in use, now.

File: lib/faraday.rb (hunk 5 above)
Comment (iMacTia) [reply lib/faraday.rb:173]: I searched for it but couldn't find any reference

**Comment by iMacTia** [comment]
Thanks for the quick review and the suggestions @olleolleolle 🙌!
Yes `middleware_mutex` is still used to ensure the middleware stack is thread-safe 👍

**Review by olleolleolle — APPROVED** [review]
This looks ready!

**Pull request merged**